Build fails when using GCC 12

CBMC build fails when using GCC 12 (latest version of GCC released in August 2022) on builds not using `-O` directives (meaning `-O0`).

Multiple new `return-type` warnings are produced when building with GCC 12.
This makes the build to fail as we use -Werror.

```
cbmc/src/util/interval.cpp: In static member function 'static exprt constant_interval_exprt::get_extreme(std::vector<exprt>, bool)':
cbmc/src/util/interval.cpp:567:1: error: control reaches end of non-void function [-Werror=return-type]
  567 | }
      | ^
cc1plus: all warnings being treated as errors
```

```
cbmc/src/util/simplify_expr_int.cpp: In member function 'simplify_exprt::resultt<> simplify_exprt::simplify_inequality_both_constant(const binary_relation_exprt&)':
cbmc/src/util/simplify_expr_int.cpp:1533:1: error: control reaches end of non-void function [-Werror=return-type]
 1533 | }
      | ^
```

It is possible that other files belonging to other submodules also fail, but the build failure stops the build earlier.

When building with CMake, it is possible to reproduce this by using `Debug` as `CMAKE_BUILD_TYPE` (`cmake -DCMAKE_BUILD_TYPE=Debug ..`).

Note that the error is not related with the addition of `-DNDEBUG`, but appears **ONLY** when `-O0` is passed (or no `-O` directives are passed).

CBMC version: 5.80.0
Operating system: Ubuntu 22.04, MacOS, Fedora
Exact command line resulting in the issue: make
What behaviour did you expect: build succeed
What happened instead: build failure

--- a/regression/invariants/driver.cpp
+++ b/regression/invariants/driver.cpp
@@ -112,6 +112,8 @@ int main(int argc, char** argv)
     UNREACHABLE_STRUCTURED(structured_error_testt, 1, "Structured error"); // NOLINT
   else if(arg=="unreachable-string")
     UNREACHABLE;
+  else if(arg == "unreachable-because")
+    UNREACHABLE_BECAUSE("Unreachable with explanation text");
   else if(arg=="data-invariant-structured")
     DATA_INVARIANT_STRUCTURED(false, structured_error_testt, 1, "Structured error"); // NOLINT
   else if(arg=="data-invariant-string")

--- a/regression/invariants/invariant-failure21/test.desc
+++ b/regression/invariants/invariant-failure21/test.desc
@@ -1,0 +1,12 @@
+CORE
+dummy_parameter.c
+unreachable-because
+^EXIT=(0|127|134|137)$
+^SIGNAL=0$
+--- begin invariant violation report ---
+Unreachable with explanation text
+Invariant check failed
+^(Backtrace)|(Backtraces not supported)$
+--
+^warning: ignoring
+^VERIFICATION SUCCESSFUL$

--- a/src/solvers/smt2/smt2_conv.cpp
+++ b/src/solvers/smt2/smt2_conv.cpp
@@ -516,10 +516,9 @@ constant_exprt smt2_convt::parse_literal(
     return from_integer(value, type);
   }
   else
-    INVARIANT(
-      false,
+    UNREACHABLE_BECAUSE(
       "smt2_convt::parse_literal should not be of unsupported type " +
-        type.id_string());
+      type.id_string());
 }
 
 exprt smt2_convt::parse_array(

--- a/src/solvers/strings/string_format_builtin_function.cpp
+++ b/src/solvers/strings/string_format_builtin_function.cpp
@@ -232,7 +232,8 @@ add_axioms_for_format_specifier(
   }
   }
 
-  INVARIANT(false, "format specifier must belong to [bBhHsScCdoxXeEfgGaAtT%n]");
+  UNREACHABLE_BECAUSE(
+    "format specifier must belong to [bBhHsScCdoxXeEfgGaAtT%n]");
 }
 
 /// Deserialize an argument for format from \p string.

--- a/src/util/invariant.h
+++ b/src/util/invariant.h
@@ -509,6 +509,12 @@ CBMC_NORETURN void report_invariant_failure(
       INVARIANT(false, "Unreachable");                                         \
       __builtin_unreachable();                                                 \
     } while(false)
+#  define UNREACHABLE_BECAUSE(REASON)                                          \
+    do                                                                         \
+    {                                                                          \
+      INVARIANT(false, REASON);                                                \
+      __builtin_unreachable();                                                 \
+    } while(false)
 #  define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                \
     do                                                                         \
     {                                                                          \
@@ -517,6 +523,7 @@ CBMC_NORETURN void report_invariant_failure(
     } while(false)
 #else
 #  define UNREACHABLE INVARIANT(false, "Unreachable")
+#  define UNREACHABLE_BECAUSE(REASON) INVARIANT(false, REASON)
 #  define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                \
     EXPAND_MACRO(INVARIANT_STRUCTURED(false, TYPENAME, __VA_ARGS__))
 #endif

--- a/src/util/invariant.h
+++ b/src/util/invariant.h
@@ -531,8 +531,17 @@ CBMC_NORETURN void report_invariant_failure(
 #define DATA_INVARIANT_STRUCTURED(CONDITION, TYPENAME, ...)                    \
   EXPAND_MACRO(INVARIANT_STRUCTURED(CONDITION, TYPENAME, __VA_ARGS__))
 
-#define UNIMPLEMENTED_FEATURE(FEATURE)                                         \
-  INVARIANT(false, std::string{"Reached unimplemented "} + (FEATURE))
+#if __GNUC__
+#  define UNIMPLEMENTED_FEATURE(FEATURE)                                       \
+    do                                                                         \
+    {                                                                          \
+      INVARIANT(false, std::string{"Reached unimplemented "} + (FEATURE));     \
+      __builtin_unreachable();                                                 \
+    } while(false)
+#else
+#  define UNIMPLEMENTED_FEATURE(FEATURE)                                       \
+    INVARIANT(false, std::string{"Reached unimplemented "} + (FEATURE))
+#endif
 
 // Legacy annotations
 

--- a/src/util/invariant.h
+++ b/src/util/invariant.h
@@ -500,9 +500,26 @@ CBMC_NORETURN void report_invariant_failure(
   EXPAND_MACRO(INVARIANT_STRUCTURED(CONDITION, TYPENAME, __VA_ARGS__))
 
 /// This should be used to mark dead code
-#define UNREACHABLE INVARIANT(false, "Unreachable")
-#define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                  \
-  EXPAND_MACRO(INVARIANT_STRUCTURED(false, TYPENAME, __VA_ARGS__))
+#ifdef __GNUC__
+// GCC 12 with -O0 fails reporting missing return when using UNREACHABLE.
+// Using __builtin_unreachable fixes this without breaking the invariant.
+#  define UNREACHABLE                                                          \
+    do                                                                         \
+    {                                                                          \
+      INVARIANT(false, "Unreachable");                                         \
+      __builtin_unreachable();                                                 \
+    } while(false)
+#  define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                \
+    do                                                                         \
+    {                                                                          \
+      EXPAND_MACRO(INVARIANT_STRUCTURED(false, TYPENAME, __VA_ARGS__));        \
+      __builtin_unreachable();                                                 \
+    } while(false)
+#else
+#  define UNREACHABLE INVARIANT(false, "Unreachable")
+#  define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                \
+    EXPAND_MACRO(INVARIANT_STRUCTURED(false, TYPENAME, __VA_ARGS__))
+#endif
 
 /// This condition should be used to document that assumptions that are
 /// made on goto_functions, goto_programs, exprts, etc. being well formed.

